### PR TITLE
Block per key-value checksum

### PR DESCRIPTION
**Pull request by cbi42**
Summary: add option `block_protection_bytes_per_key` and implementation for block per key-value checksum. The main changes are
1. checksum construction and verification in block.cc/h
2. pass the option `block_protection_bytes_per_key` around (mainly for methods defined in table_cache.h)
3. unit tests/crash test updates

Tests:
* Added unit tests
* Crash test: `python3 tools/db_crashtest.py blackbox --simple --block_protection_bytes_per_key=1 --write_buffer_size=1048576`

Follow up (maybe as a separate PR): make sure corruption status returned from BlockIters are correctly handled.

Performance: 
Turning on block per KV protection has a non-trivial negative impact on read performance and costs additional memory.
For memory, each block includes additional 24 bytes for checksum-related states beside checksum itself. For CPU, I set up a DB of size ~1.2GB with 5M keys (32 bytes key and 200 bytes value) which compacts to ~5 SST files (target file size 256 MB) in L6 without compression. I tested readrandom performance with various block cache size (to mimic various cache hit rates):

```
SETUP
make OPTIMIZE_LEVEL="-O3" USE_LTO=1 DEBUG_LEVEL=0 -j32 db_bench
./db_bench -benchmarks=fillseq,compact0,waitforcompaction,compact,waitforcompaction -write_buffer_size=33554432 -level_compaction_dynamic_level_bytes=true -max_background_jobs=8 -target_file_size_base=268435456 --num=5000000 --key_size=32 --value_size=200 --compression_type=none

BENCHMARK
./db_bench --use_existing_db -benchmarks=readtocache,readrandom[-X10] --num=5000000 --key_size=32 --disable_auto_compactions --reads=1000000 --block_protection_bytes_per_key=[0|1] --cache_size=$CACHESIZE

The readrandom ops/sec looks like the following:
Block cache size:  2GB        1.2GB * 0.9    1.2GB * 0.8     1.2GB * 0.5   8MB
Main              240805     223604         198176           161653       139040
PR prot_bytes=0   238691     226693         200127           161082       141153
PR prot_bytes=1   214983     193199         178532           137013       108211
prot_bytes=1 vs    -10%        -15%          -10.8%          -15%        -23%
prot_bytes=0
```

The benchmark has a lot of variance, but there was a 5% to 25% regression in this benchmark with different cache hit rates.

### CODE REVIEW

**Review by ajkr — COMMENTED** [review]
Thanks for the complete test coverage and complete argument passing (that is, no default values for the new parameters)

I have a few files remaining to review

**Review by ajkr — APPROVED** [review]
LGTM! Congrats on finishing this huge feature

File: table/block_based/block.cc [hunk 1]
```diff
@@ -381,14 +396,26 @@ bool DataBlockIter::SeekForGetImpl(const Slice& target) {
...
+    // in SeekForGet().
+    TEST_SYNC_POINT_CALLBACK("DataBlockIter::SeekForGetImpl::value",
+                             (void*)value_.data());
+    if (!Block::VerifyChecksum(
+            protection_bytes_per_key_,
+            kv_checksum_ + protection_bytes_per_key_ * cur_entry_idx_,
+            raw_key_.GetKey(), value_)) {
+      PerKVChecksumCorruptionError();
+      return true;
+    }
```
[the remark this answers is not in the document]
Comment (cbi42) [reply table/block_based/block.cc]: They are only needed in some cases as discussed offline (e.g., when non_shared=0). Removed them for now.

**Review by cbi42 — COMMENTED** [review]
Thanks for the review! I've update the PR summary with new benchmark results, should be good to review again.

File: db/version_edit_handler.cc [hunk 2]
```diff
@@ -564,12 +564,13 @@ Status VersionEditHandler::LoadTables(ColumnFamilyData* cfd,
   assert(builder_iter->second != nullptr);
   VersionBuilder* builder = builder_iter->second->version_builder();
   assert(builder);
+  const MutableCFOptions* moptions = cfd->GetLatestMutableCFOptions();
   Status s = builder->LoadTableHandlers(
```
Comment (hx235) [inline db/version_edit_handler.cc:568]: Just so you know : potential merge conflict with https://github.com/facebook/rocksdb/pull/11288

File: table/block_based/block.cc [hunk 3]
```diff
@@ -1125,6 +1284,7 @@ size_t Block::ApproximateMemoryUsage() const {
   if (read_amp_bitmap_) {
     usage += read_amp_bitmap_->ApproximateMemoryUsage();
   }
+  usage += checksum_size_;
```
Comment (ajkr) [inline table/block_based/block.cc:1287]: Thanks for remembering this

**Review by ajkr — APPROVED** [review]
LGTM again

**Pull request closed without merging**